kratos/solving_strategies/schemes/residual_based_bdf_displacement_scheme.h

```diff
@@ -183,74 +267,36 @@ class ResidualBasedBDFDisplacementScheme
+            if (velpos > -1) {
+                for (std::size_t i_dim = 0; i_dim < dimension; ++i_dim) {
+                    if (it_node->GetDof(vel_components[i_dim], velpos + i_dim).IsFixed() && !predicted[i_dim]) {
+                        un0[i_dim] = dotun0[i_dim];
+                        for (std::size_t i_order = 1; i_order < BDFBaseType::mOrder + 1; ++i_order)
+                            un0[i_dim] -= BDFBaseType::mBDF[i_order] * it_node->FastGetSolutionStepValue(disp_components[i_dim], i_order);
+                        un0[i_dim] /= BDFBaseType::mBDF[i_dim];
+                        predicted[i_dim] = true;
+                    }
                 }
             }
-
-            for (std::size_t i_order = 2; i_order < BDFBaseType::mOrder + 1; ++i_order) {
-                const array_1d<double, 3>& dotun = it_node->FastGetSolutionStepValue(VELOCITY,     i_order);
-                const array_1d<double, 3>& un = it_node->FastGetSolutionStepValue(DISPLACEMENT, i_order);
-
-                if (it_node->HasDofFor(ACCELERATION_X)) {
-                    if (it_node -> IsFixed(ACCELERATION_X)) {
-                        dotun0[0] -= (BDFBaseType::mBDF[i_order] * dotun[0])/BDFBaseType::mBDF[0];
-                        un0[0] -= (BDFBaseType::mBDF[i_order] * un[0])/BDFBaseType::mBDF[0];
-                } } else if (it_node->HasDofFor(VELOCITY_X)) {
-                    if (it_node -> IsFixed(VELOCITY_X)) {
-                        un0[0] -= (BDFBaseType::mBDF[i_order] * un[0])/BDFBaseType::mBDF[0];
-                } }
-
-                if (it_node->HasDofFor(ACCELERATION_Y)) {
-                    if (it_node -> IsFixed(ACCELERATION_Y)) {
-                        dotun0[1] -= (BDFBaseType::mBDF[i_order] * dotun[1])/BDFBaseType::mBDF[0];
-                        un0[1] -= (BDFBaseType::mBDF[i_order] * un[1])/BDFBaseType::mBDF[0];
-                } } else if (it_node->HasDofFor(VELOCITY_Y)) {
-                    if (it_node -> IsFixed(VELOCITY_X)) {
-                        un0[1] -= (BDFBaseType::mBDF[i_order] * un[1])/BDFBaseType::mBDF[0];
-                } }
-
-                // For 3D cases
-                if (it_node -> HasDofFor(DISPLACEMENT_Z)) {
-                    if (it_node->HasDofFor(ACCELERATION_Z)) {
-                        if (it_node -> IsFixed(ACCELERATION_Z)) {
-                            dotun0[1] -= (BDFBaseType::mBDF[i_order] * dotun[2])/BDFBaseType::mBDF[0];
-                            un0[1] -= (BDFBaseType::mBDF[i_order] * un[2])/BDFBaseType::mBDF[0];
-                    } } else if (it_node->HasDofFor(VELOCITY_Y)) {
-                        if (it_node -> IsFixed(VELOCITY_X)) {
-                            un0[1] -= (BDFBaseType::mBDF[i_order] * un[2])/BDFBaseType::mBDF[0];
-                    } }
+            for (std::size_t i_dim = 0; i_dim < dimension; ++i_dim) {
+                if (!it_node->GetDof(disp_components[i_dim], disppos + i_dim).IsFixed() && !predicted[i_dim]) {
```

Comment:
change the order in the if. I mean

~~~
if ( !predicted[i_dim] && !it_node->GetDof(disp_components[i_dim], disppos + i_dim).IsFixed())
~~~

i suggest this since GetDof is anyway more expensive than checking  !predicted[i_dim] so you will avoid the expensive operation if the first check is false

Reply:
btw, are u sure this works with variable time step?

Reply:
Yes, it does